### ISSUE TITLE
jsx-uses-vars breaks eslint prefer-const

### ISSUE THREAD

**Issue by lukeapage** [yannickcr/eslint-plugin-react#716]
I have code like this
```jsx
function X(props) {
    const { filterList } = props;

    let filters = _.chain(filterList)
        .map((filter, index) =>
            <Z/>
        )
        .value();

    return (
        <div>
            {filters}
        </div>
    );
}
```
but with prefer-const on, the prefer const error on `filters` is removed when I switch on `jsx-uses-vars`

**Comment by lukeapage** [comment]
I created a repro repo..
https://github.com/lukeapage/eslint-react-bug-prefer-const-716

**Comment by yannickcr** [comment]
Thanks for the issue, but if I'm right there is nothing we can do on our side to prevent this, since the issue seems to come from ESLint itself (https://github.com/eslint/eslint/issues/5837), right?

**Comment by ljharb** [comment]
It sounds like with `prefer-const` there's a (correct) error, and with `prefer-const` + `jsx-uses-vars`, there's (incorrectly) not an error.

**Comment by yannickcr** [comment]
`prefer-const` is ignoring variables with `eslintUsed` set on it. Since that's how we mark a variable as used in ESLint I don't see what we can do here 😐 

But:

* We mark `JSXOpeningElement` (`<App />`) and `JSXExpressionContainer` (`{filters}`) as used but ESLint seems to correctly handle `JSXExpressionContainer` without the help of this rule. Maybe we could remove this? (All tests are still passing without it, but maybe we're missing a pattern?)

This will solve the case described by @lukeapage but we'll still have the problem with `JSXOpeningElement`:

```jsx
let App = <div />; // should warn "'App' is never reassigned. Use 'const' instead" but will not
<App />
```

* Apparently they are not sure if this check on `eslintUsed` is still useful (https://github.com/eslint/eslint/issues/5837#issuecomment-235490801), so maybe they will remove it? In this case it would solve the entire problem.

**Comment by lukeapage** [comment]
Yeah sorry for not updating after tracking down.

I left this open until a solution is found, in case something needs to be done here as welll.

**Comment by Wilfred** [comment]
I hit this today. A minimal repro:

```jsx
function foo() {
  let iCouldBeConst = 6;
  return <th colSpan={iCouldBeConst}></th>;
}
```

and eslint config:

```json
{
    "parserOptions": {
        "ecmaVersion": 6,
        "ecmaFeatures": {
            "jsx": true
        },
        "sourceType": "module"
    },
    "rules": {
        "react/jsx-uses-vars": 2,
        "prefer-const": 2
    },
    "plugins": [
        "react"
    ]
}
```

Sounds like removing `eslintUsed` on `JSXExpressionContainer` would be a definite improvement, since that would fix both my situation and OP.

**Comment by yannickcr** [comment]
> Sounds like removing eslintUsed on JSXExpressionContainer would be a definite improvement, since that would fix both my situation and OP.

Yes, I will remove this. It does not fix the whole issue with `prefer-const` but it's better than nothing for now.

**Comment by yannickcr** [comment]
Keep it open for now due to https://github.com/yannickcr/eslint-plugin-react/issues/716#issuecomment-235762191

**Comment by Wilfred** [comment]
Thanks :)

**Comment by yannickcr** [comment]
Fixed in [ESLint 3.4.0](http://eslint.org/blog/2016/08/eslint-v3.4.0-released#bug-fixes)
